service/patch.go

```diff
@@ -104,36 +104,48 @@ func (uis *UIServer) patchPage(w http.ResponseWriter, r *http.Request) {
 		"base", "patch_version.html", "base_angular.html", "menu.html")
 }
 
-func (uis *UIServer) schedulePatch(w http.ResponseWriter, r *http.Request) {
+func (uis *UIServer) schedulePatchUI(w http.ResponseWriter, r *http.Request) {
 	projCtx := MustHaveProjectContext(r)
 	if projCtx.Patch == nil {
-		http.Error(w, "patch not found", http.StatusNotFound)
-		return
+		uis.LoggedError(w, r, http.StatusNotFound, errors.New("patch not found"))
 	}
 	curUser := gimlet.GetUser(r.Context())
 	if curUser == nil {
-		http.Error(w, "Not authorized to schedule patch", http.StatusUnauthorized)
+		uis.LoggedError(w, r, http.StatusUnauthorized, errors.New("Not authorized to schedule patch"))
+	}
+	patchUpdateReq := patchVariantsTasksRequest{}
+	if err := util.ReadJSONInto(util.NewRequestReader(r), &patchUpdateReq); err != nil {
+		uis.LoggedError(w, r, http.StatusBadRequest, err)
+	}
+
+	err, status, successMessage, versionId := uis.SchedulePatch(r.Context(), projCtx, patchUpdateReq)
+	if err != nil {
+		uis.LoggedError(w, r, status, err)
 		return
 	}
-	// grab patch again, as the diff  was excluded
+
+	PushFlash(uis.CookieStore, r, w, NewSuccessFlash(successMessage))
+	gimlet.WriteJSON(w, struct {
+		VersionId string `json:"version"`
+	}{versionId})
+
+}
+
+// SchedulePatch schedules a patch, returning an error, an http status code,
+// an optional success message, and an optional version ID.
+func (uis *UIServer) SchedulePatch(ctx context.Context, projCtx projectContext, patchUpdateReq patchVariantsTasksRequest) (error, int, string, string) {
```

Comment:
This isn't usable from GQL because the service package imports the GQL package, and we don't have access to the UI server there. This should be written in a way that it makes sense to put in lifecycle.go

Reply:
Done.

Reply:
It still needs to be in a different package. Service imports graphql

Reply:
Done.